### PR TITLE
Put S-expressions in their own library

### PR DESCRIPTION
**Pull request by mariari**

### CODE REVIEW

File: library/Sexp/test/Main.hs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import qualified Sexp
+import qualified Sexp.Parser
+import qualified Sexp.SimplifiedPasses
+import qualified Test.Tasty as T
+
+allCheckedTests :: T.TestTree
+allCheckedTests =
+  T.testGroup
+    "All tests that are checked"
```
Comment (rokopt) [inline library/Sexp/test/Main.hs:12]: Trivial comment here, and I know this is taken from existing code, but `"All tests that are checked"` sounds odd and non-specific to me.  Would something like simply `"S-expression library tests"` read better?

File: doc/Code/Library.org [hunk 2]
```diff
@@ -1467,6 +1467,113 @@ Can be added to via core translation
...
+  + [[Parser/Internal]]
+  + [[Sexp/Types]]
+***** Types <<Sexp/Types>>
+- _Relies on_
+  + [[Library]]
+  + [[LineNum]]
+  + [[NameSymbol]]
+***** Structure
+****** Frontend <<Structure/Frontend>>
+Hardeing S-expressions into a more readable form. Here we use a
```
Comment (rokopt) [inline doc/Code/Library.org:1497]: Hardeing -> Hardening

File: doc/Code/Library.org [hunk 3]
```diff
@@ -1467,6 +1467,113 @@ Can be added to via core translation
...
+mixture of record structures and aliases. Each cover a form that we
+wish to talk about rather than just match away at
+- _The form for transformation follows this structure_
+#+begin_src haskell
+  -- the data type
+  data Form = ... deriving (Show)
+  is<Form>   :: Sexp.T -> Bool
+  to<Form>   :: Sexp.T -> Maybe <Form>
+  from<Form> :: <Form> -> Sexp.T
```
Comment (rokopt) [inline doc/Code/Library.org:1506]: I'm guessing there are invariants that we might want to note here, such as:

```haskell
`ForAll s : SExp.T, is<Form> s = True <-> isJust (to<Form> s)`

`to<Form> . from<Form> = Just`
```

File: doc/Code/Library.org [hunk 4]
```diff
@@ -1467,6 +1467,113 @@ Can be added to via core translation
...
+ 1. Figure out if we can even express a spec system in
+    Haskell... =to<Form>= and =From<From>= have the exact same signature
+ 2. replace the repeat code with the =to<Form>= with an abstraction
+ 3. put the meta data with the form so we don't have to do it by
+    hand in the code that uses this
+    1. Use =Juvix.Library.LineNum=
+    2. append the =Form= With this
+    3. have to<Form> fill this
+    4. Have extra smart consturctors that are =<form>=, so that we
+       can automatically fill in this meta data
```
Comment (rokopt) [inline doc/Code/Library.org:1554]: This comment looks like it's repeated between `Frontend` and `Transition` -- should it be in a single place?

File: library/Sexp/src/Juvix/Sexp/Structure/generator.lisp [hunk 5]
```diff
@@ -263,71 +263,73 @@ and a rhs that may contain a guard, so no = is assumed for the rhs"
                        (body from-body))
                   "")))))))
 
-(generate-haskell "ArgBody" '("sexp" "sexp") nil)
+(defun frontend-types ()
+  (generate-haskell "Type" (repeat 3 "sexp") "type" :list-star t)
```
Comment (rokopt) [inline library/Sexp/src/Juvix/Sexp/Structure/generator.lisp:267]: If I'm reading this file correctly, a bunch of spaces were added at the beginnings of lines, and some of the definitions were re-ordered.  Is any of that necessary, or is this some kind of reformatting accident?  Or is the file auto-generated and some structure changes made it reformat/reorder everything?

File: library/Sexp/src/Juvix/Sexp/Structure/Transition.hs [hunk 6]
```diff
@@ -0,0 +1,345 @@
+{-# LANGUAGE DeriveTraversable #-}
```
Comment (rokopt) [inline library/Sexp/src/Juvix/Sexp/Structure/Transition.hs:1]: Nice, I like how splitting it up into two files makes it read.

**Review by rokopt — APPROVED** [review]

File: library/Sexp/src/Juvix/Sexp/Structure/generator.lisp (hunk 5 above)
Comment (mariari) [reply library/Sexp/src/Juvix/Sexp/Structure/generator.lisp:267]: I made functions, and split it between a few files!

Only 2 forms differ both the handlers which were incorrect

File: doc/Code/Library.org (hunk 4 above)
Comment (mariari) [reply doc/Code/Library.org:1554]: perhaps but for now it should stay like this I think?

**Pull request merged**